src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs

```diff
@@ -283,5 +283,44 @@ private static Conversion ToConversion(OverloadResolutionResult<MethodSymbol> re
 
             return new Conversion(ConversionKind.MethodGroup, method, methodGroup.IsExtensionMethodGroup);
         }
+
+        public override Conversion GetStackAllocConversion(BoundStackAllocArrayCreation sourceExpression, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            Debug.Assert((object)sourceExpression.Type == null);
+            Debug.Assert(sourceExpression.ElementType != null);
+
+            var pointerConversion = default(Conversion);
+
+            if (sourceExpression.Syntax.IsVariableDeclarationInitialization())
+            {
+                var sourceAsPointer = new PointerTypeSymbol(sourceExpression.ElementType);
+                pointerConversion = ClassifyImplicitConversionFromType(sourceAsPointer, destination, ref useSiteDiagnostics);
+            }
+
+            if (pointerConversion.IsValid)
+            {
+                // Report unsafe errors
+                _binder.ReportUnsafeIfNotAllowed(sourceExpression.Syntax.Location, ref useSiteDiagnostics);
+
+                return Conversion.MakeStackAllocToPointerType(pointerConversion);
+            }
+            else
+            {
+                var spanType = _binder.GetWellKnownType(WellKnownType.System_Span_T, ref useSiteDiagnostics).Construct(sourceExpression.ElementType);
+                var spanConversion = ClassifyImplicitConversionFromType(spanType, destination, ref useSiteDiagnostics);
+
+                if (spanConversion.Exists)
+                {
+                    // Report errors if Span ctor is missing, or using an older C# version
+                    Binder.CheckFeatureAvailability(sourceExpression.Syntax, MessageID.IDS_FeatureRefStructs, ref useSiteDiagnostics);
+                    Binder.GetWellKnownTypeMember(_binder.Compilation, WellKnownMember.System_Span__ctor, out DiagnosticInfo memberDiagnosticInfo);
+                    HashSetExtensions.InitializeAndAdd(ref useSiteDiagnostics, memberDiagnosticInfo);
+
+                    return Conversion.MakeStackAllocToSpanType(spanConversion);
+                }
+            }
+
+            return Conversion.NoConversion;
```

Comment:
Can you please explain? if `pointerConversion` was not valid, this means it will contain the pointer conversion, but an unsuccessful one, which is not what we want. In all cases, this function should return a stackalloc conversion, not a pointer conversion that we synthesized here.

---
In reply to: [132315778](https://github.com/dotnet/roslyn/pull/21006#discussion_r132315778) [](ancestors = 132315778)